File: java-samples/poslink/EposPosLinkExample/src/main/java/com/example/eposposlinkexample/teya/TransactionRecorder.java
package com.example.eposposlinkexample.teya;

import com.example.eposposlinkexample.models.TransactionRecord;

import com.teya.unifiedepossdk.PaymentState;
import com.teya.unifiedepossdk.PaymentStateDetails;
import com.teya.unifiedepossdk.RefundResult;
import com.teya.unifiedepossdk.RefundResultDetails;
import com.teya.unifiedepossdk.models.GatewayPaymentId;
import com.teya.unifiedepossdk.models.GatewayRefundId;
import com.teya.unifiedepossdk.models.TransactionType;

import java.util.UUID;

final class TransactionRecorder {

    private TransactionRecorder() {
    }

    static void recordPaymentIfFinal(PaymentStateDetails state, TransactionType type, boolean tabPayment,
                                     boolean cashPayment) {
        if (!state.isFinal()) {
            return;
        }
        GatewayPaymentId gatewayPaymentId = state.getGatewayPaymentId();
        Long timestamp = state.getTransactionTimestamp();
        TransactionStore.getInstance().upsert(new TransactionRecord(
                state.getEposTransactionId(),
                type,
                state.getState() == PaymentState.Successful,
                state.getAmount(),
                state.getCurrency(),
                gatewayPaymentId != null ? gatewayPaymentId.getId() : null,
                timestamp != null ? timestamp : System.currentTimeMillis(),
                false,
                cashPayment,
                tabPayment));
    }

    static void recordRefund(RefundResultDetails refundResult, String gatewayPaymentId, int amountMinor,
                             String currency) {
        GatewayRefundId gatewayRefundId = refundResult.getGatewayRefundId();
        TransactionStore.getInstance().upsert(new TransactionRecord(
                gatewayRefundId != null ? gatewayRefundId.getId() : UUID.randomUUID().toString(),
                TransactionType.Refund,
                refundResult.getResult() == RefundResult.Success,
                amountMinor,
                currency,
                null,
                System.currentTimeMillis(),
                false,
                false,
                false));

        if (refundResult.getResult() == RefundResult.Success) {
            TransactionStore.getInstance().markRefunded(gatewayPaymentId);
        }
    }
}

File: android-samples/poslink/EposPosLinkExample/app/src/main/java/com/example/eposappexample/poslink/transactions/TransactionRecorder.kt
package com.example.eposappexample.poslink.transactions

import com.teya.unifiedepossdk.PaymentState
import com.teya.unifiedepossdk.PaymentStateDetails
import com.teya.unifiedepossdk.RefundResult
import com.teya.unifiedepossdk.RefundResultDetails
import com.teya.unifiedepossdk.models.TransactionType
import java.util.UUID

object TransactionRecorder {

    fun recordPaymentIfFinal(
        state: PaymentStateDetails,
        type: TransactionType,
        isTab: Boolean = false,
        isCash: Boolean = false,
    ) {
        if (!state.isFinal) return
        TransactionStore.upsert(
            TransactionRecord(
                id = state.eposTransactionId,
                type = type,
                isSuccess = state.state == PaymentState.Successful,
                amountMinor = state.amount,
                currency = state.currency,
                gatewayPaymentId = state.gatewayPaymentId?.id,
                timestamp = state.transactionTimestamp ?: System.currentTimeMillis(),
                isCashPayment = isCash,
                isTabPayment = isTab,
            )
        )
    }

    fun recordRefund(
        refundResult: RefundResultDetails,
        gatewayPaymentId: String,
        amountMinor: Int,
        currency: String,
    ) {
        TransactionStore.upsert(
            TransactionRecord(
                id = refundResult.gatewayRefundId?.id ?: UUID.randomUUID().toString(),
                type = TransactionType.Refund,
                isSuccess = refundResult.result == RefundResult.Success,
                amountMinor = amountMinor,
                currency = currency,
                gatewayPaymentId = null,
                timestamp = System.currentTimeMillis(),
            )
        )
        if (refundResult.result == RefundResult.Success) {
            TransactionStore.markRefunded(gatewayPaymentId)
        }
    }
}